### PR TITLE
ci: release automation for unsigned artifacts (stage 1)

### PR DESCRIPTION
**Pull request by v1v**
## Summary

Automate the release based on a tag, for such it creates the binary and then upload it to a GitHub release using the [electro-build](https://www.electron.build/configuration/publish#githuboptions) approach

## Implementation details

0) Renamed the product name without spaces as those are the assets published already in the release 
1) Configure the GitHub token context
2) Run the release command as explained in the `RELEASE.md` file

## Follow ups

1) TBD - Transition the release from draft to ready (likely to be configured in electron-build -> 
> releaseType = draft “draft” | “prerelease” | “release” | “undefined” - The type of release. By default draft release will be created.
2) Signing

### CODE REVIEW

File: .ci/Jenkinsfile
Comment (v1v) [inline .ci/Jenkinsfile:16]: TBD

File: RELEASE.md
Comment (v1v) [inline RELEASE.md:23]: If needed, it could be done automatically in the CI with an input parameter, for instance -> [apm-agent-rum-js](https://github.com/elastic/apm-agent-rum-js/blob/main/RELEASE.md#ci-based) uses an input approval approach and it's UI based.

File: package.json
Comment (v1v) [inline package.json:45]: Align the existing approach done in the releases, see https://github.com/elastic/synthetics-recorder/releases/tag/untagged-f5efbb08222e46cedb8f

**Comment by v1v** [comment]
@justinkambic and @kyungeunni, I need some help to understand what's the below error about:

```
[2022-03-31T12:02:17.965Z] > cp -r ./electron/. ./build
[2022-03-31T12:02:17.965Z] 
[2022-03-31T12:02:17.965Z]   • electron-builder  version=23.0.3 os=5.13.0-1021-gcp
[2022-03-31T12:02:17.965Z]   • loaded configuration  file=package.json ("build" field)
[2022-03-31T12:02:17.965Z]   ⨯ /var/lib/jenkins/workspace/m_synthetics-recorder-mbp_PR-183/src/github.com/elastic/synthetics-recorder/node_modules/iconv-corefoundation/lib/native.node: invalid ELF header  failedTask=build stackTrace=Error: /var/lib/jenkins/workspace/m_synthetics-recorder-mbp_PR-183/src/github.com/elastic/synthetics-recorder/node_modules/iconv-corefoundation/lib/native.node: invalid ELF header
[2022-03-31T12:02:17.965Z]     at Object.Module._extensions..node (internal/modules/cjs/loader.js:1131:18)
[2022-03-31T12:02:17.965Z]     at Module.load (internal/modules/cjs/loader.js:937:32)
[2022-03-31T12:02:17.965Z]     at Function.Module._load (internal/modules/cjs/loader.js:778:12)
[2022-03-31T12:02:17.965Z]     at Module.require (internal/modules/cjs/loader.js:961:19)
[2022-03-31T12:02:17.965Z]     at require (internal/modules/cjs/helpers.js:92:18)
[2022-03-31T12:02:17.965Z]     at Object.<anonymous> (/var/lib/jenkins/workspace/m_synthetics-recorder-mbp_PR-183/src/github.com/elastic/synthetics-recorder/node_modules/iconv-corefoundation/lib/native.js:5:18)
[2022-03-31T12:02:17.965Z]     at Module._compile (internal/modules/cjs/loader.js:1072:14)
[2022-03-31T12:02:17.965Z]     at Object.Module._extensions..js (internal/modules/cjs/loader.js:1101:10)
[2022-03-31T12:02:17.965Z]     at Module.load (internal/modules/cjs/loader.js:937:32)
[2022-03-31T12:02:17.965Z]     at Function.Module._load (internal/modules/cjs/loader.js:778:12)
[2022-03-31T12:02:17.965Z]     at Module.require (internal/modules/cjs/loader.js:961:19)
[2022-03-31T12:02:17.965Z]     at require (internal/modules/cjs/helpers.js:92:18)
[2022-03-31T12:02:17.965Z]     at Object.<anonymous> (/var/lib/jenkins/workspace/m_synthetics-recorder-mbp_PR-183/src/github.com/elastic/synthetics-recorder/node_modules/iconv-corefoundation/lib/index.js:6:18)
[2022-03-31T12:02:17.965Z]     at Module._compile (internal/modules/cjs/loader.js:1072:14)
[2022-03-31T12:02:17.965Z]     at Object.Module._extensions..js (internal/modules/cjs/loader.js:1101:10)
[2022-03-31T12:02:17.965Z]     at Module.load (internal/modules/cjs/loader.js:937:32)
[2022-03-31T12:02:17.965Z]     at Function.Module._load (internal/modules/cjs/loader.js:778:12)
[2022-03-31T12:02:17.965Z]     at Module.require (internal/modules/cjs/loader.js:961:19)
[2022-03-31T12:02:17.965Z]     at require (internal/modules/cjs/helpers.js:92:18)
[2022-03-31T12:02:17.965Z]     at Object.<anonymous> (/var/lib/jenkins/workspace/m_synthetics-recorder-mbp_PR-183/src/github.com/elastic/synthetics-recorder/node_modules/dmg-license/lib/Labels.js:3:32)
[2022-03-31T12:02:17.965Z]     at Module._compile (internal/modules/cjs/loader.js:1072:14)
[2022-03-31T12:02:17.965Z]     at Object.Module._extensions..js (internal/modules/cjs/loader.js:1101:10)
[2022-03-31T12:02:17.965Z] npm ERR! code ELIFECYCLE
[2022-03-31T12:02:17.965Z] npm ERR! errno 1
[2022-03-31T12:02:17.965Z] npm ERR! synthetics-recorder@0.0.1-alpha.2 release: `npm run build && electron-builder -mwl --publish=always`
[2022-03-31T12:02:17.965Z] npm ERR! Exit status 1
[2022-03-31T12:02:17.965Z] npm ERR! 
[2022-03-31T12:02:17.965Z] npm ERR! Failed at the synthetics-recorder@0.0.1-alpha.2 release script.
[2022-03-31T12:02:17.965Z] npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
[2022-03-31T12:02:17.965Z] 
[2022-03-31T12:02:17.965Z] npm ERR! A complete log of this run can be found in:
[2022-03-31T12:02:17.965Z] npm ERR!     /var/lib/jenkins/workspace/m_synthetics-recorder-mbp_PR-183/.npm/_logs/2022-03-31T12_02_15_911Z-debug.log
```

The whole console log can be seen [here](https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-rum/pipelines/synthetics-recorder-mbp/pipelines/PR-183/runs/2/steps/274/log/?start=0)

Thanks

**Comment by justinkambic** [comment]
>  I need some help to understand what's the below error about

Hi @v1v - we are going to try to run this release script on Linux and see if we can reproduce the issue.

Also, is it possible for us to attempt to run this on a mac worker to see if the same problem occurs?

This error message is usually the result of attempting to reference some native binary that was built for a different platform, so it's possible that the step before this where we 

```
[2022-03-31T12:19:28.866Z] > cp -r ./electron/. ./build
```

is including some binary that is not intended for the OS of the worker.

The package that threw the error is a dependency of `DMG Builder`, which is one of `electron-builder`'s deps. Given that it's a mac-specific package causing the problem, we're curious if we could see if the problem still occurs when using a mac worker.

**Comment by v1v** [comment]
@justinkambic and @kyungeunni, this is much better, it works on MacOS, at least the dist generation, unfortunately the publishing in GitHub is skipped for some reason, I've no clue what's the expected behaviour with the electron-builder, please see below:

```
[2022-03-31T17:35:23.742Z]   • skipped publishing  file=Elastic-Synthetics-Recorder-0.0.1-alpha.2-mac.zip reason=existing type not compatible with publishing type tag=v0.0.1-alpha.2 version=0.0.1-alpha.2 existingType=pre-release publishingType=draft
```

See the [whole log](https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-rum/pipelines/synthetics-recorder-mbp/pipelines/PR-183/runs/6/nodes/253/log/?start=0)

My plan was to create a draft release automatically, so in theory all the bits and pieces are in place, but I don't know if I missed something, please let me know if you find anything

Thanks

**Comment by kyungeunni** [comment]
Hey @v1v, great progress 🎉 thank you so much for pushing it forward!

What I can see from this error message is that we already have artifacts published with `v0.0.1-alpha.2` tag that is in pre-release, but we tried to create a draft release with the same tag. I think we either need to delete all the artifacts pushed with that tag and try it again, or just bump the version in our package.json and try again.

Apart from the tag issue, we also need to provide `GH_TOKEN` as env variable when running the release script. You can find the documentation [here](https://github.com/elastic/synthetics-recorder/blob/main/RELEASE.md), We've been using the [personal token](https://www.google.com/url?q=https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/creating-a-personal-access-token&sa=D&source=calendar&ust=1648992317788959&usg=AOvVaw24yNSE3bL9Q6WxT9iJJF3V) with repo scope for this so far, but not sure what token is supposed to be used in Jenkins.

**Comment by v1v** [comment]
Thanks for the hint, I just https://github.com/elastic/synthetics-recorder/pull/183/commits/657b11c071d7bf6b18204d3ccf31d12bda62644e but it failed somehow, although it produced a draft PR: https://github.com/elastic/synthetics-recorder/releases/tag/untagged-9ccbf393cb6a3808c7ff with a weird name on it.

I guess, the electron-builder is not atomic and if something bad happens while running the release then it will left the release in a partial state.


The log details can be found [here](https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-rum/pipelines/synthetics-recorder-mbp/pipelines/PR-183/runs/8/nodes/233/log/?start=0)


I'll try again and see if the issue was maybe some flakiness

**Comment by kyungeunni** [comment]
> although it produced a draft PR: https://github.com/elastic/synthetics-recorder/releases/tag/untagged-9ccbf393cb6a3808c7ff with a weird name on it.

I presume electorn-builder expect to have the tag corresponding to the version in package.json by the look of the part of the url `untagged`, but not 100% sure.

I've looked at the documentation from electron builder and there are two options to pass over the note
1. use CLI -c arg to set `releaseInfo` config
2. use static file for release note, such as `release-note.md`

Here's the screenshot from the documentation. (I can't link directly to the section, but you can find releaseInfo from [this page](https://www.electron.build/configuration/configuration))
image

**Comment by v1v** [comment]
Thanks!

I assume those configuration details should be added to the `package.json` if they need to be forced. Although, if we release from a git tag then there is no need to do much since IIUC, it will work out of the box.

If so, I'll move this PR to ready for review, and we can see how it goes with the tag release

File: .ci/Jenkinsfile
Comment (v1v) [inline .ci/Jenkinsfile:8]:
```suggestion
  agent { label 'ubuntu-20 && immutable' }
```

File: .ci/Jenkinsfile
Comment (v1v) [reply .ci/Jenkinsfile:18]: @justinkambic and @kyungeunni, I need your input here, what slack channel and mailing list should we use when the release has done or failed?

File: .ci/Jenkinsfile
Comment (v1v) [inline .ci/Jenkinsfile:62]:
```suggestion
      when { expression { return !isTag() } }
```

File: .ci/Jenkinsfile
Comment (v1v) [inline .ci/Jenkinsfile:86]:
```suggestion
      when {
        beforeAgent true
        expression { return !isTag() }
      }
```

File: .ci/Jenkinsfile
Comment (v1v) [inline .ci/Jenkinsfile:110]:
```suggestion
      when { expression { return !isTag() } }
```

File: .ci/Jenkinsfile
Comment (v1v) [inline .ci/Jenkinsfile:124]:
```suggestion
      when {
```

File: .ci/Jenkinsfile
Comment (v1v) [inline .ci/Jenkinsfile:129]:
```suggestion
      }
      agent { label 'macosx' }
```

File: package.json
Comment (v1v) [inline package.json:3]:
```suggestion
  "version": "0.0.1-alpha.2",
```

File: .ci/Jenkinsfile
Comment (v1v) [inline .ci/Jenkinsfile:47]:
```suggestion
      when { expression { return !isTag() } }
```

File: package-lock.json
Comment (v1v) [inline package-lock.json:3]:
```suggestion
  "version": "0.0.1-alpha.2",
```

**Review by kuisathaverat — APPROVED** [review]

**Comment by justinkambic** [comment]
@v1v it seems that the `Release` and `Dist` steps are skipped in the most recent run, is this done on purpose?

File: .ci/Jenkinsfile
Comment (justinkambic) [reply .ci/Jenkinsfile:16]: @v1v I'll message you offline about this.

**Comment by v1v** [comment]
> @v1v it seems that the Release and Dist steps are skipped in the most recent run, is this done on purpose?

Indeed, these changes will be run only on a git tag event, so PRs or branches will not run a release stage, but tags. 

I used this PR to test the stages by disabling the tag validation, see https://github.com/elastic/synthetics-recorder/pull/183/commits/06043ab1a1351f0c12d04e48e136651e816510e0

As soon as we merge and run a release we will see if it works those stages, though I tested it already in 

image

**Comment by justinkambic** [comment]
Thanks for the quick explanation!

**Review by kyungeunni — APPROVED** [review]
LGTM! 

I noticed the comment regarding the slack channel and email is still pending. As far as I know, @justinkambic has contacted you with the details so I'm approving it for now. Let me know if you need any further input 🙏

File: .ci/Jenkinsfile
Comment (justinkambic) [reply .ci/Jenkinsfile:16]: If we update the email group with synthrum@elastic.co we will be good here.

**Review by justinkambic — APPROVED** [review]
LGTM pending email update.

File: .ci/Jenkinsfile
Comment (v1v) [inline .ci/Jenkinsfile:16]:
```suggestion
    NOTIFY_TO = 'synthrum+synthetics-recorder@elastic.co'
```

**Pull request merged**
